### ISSUE TITLE
Add a retry method

### ISSUE THREAD

**Issue by feinstein** [dart-lang/web_socket_channel#137]
`http` has a [`retry`](https://pub.dev/packages/retry) package where we can configure our retry strategy.

Many Apps, like mine, need to have a contiguous and reliable websocket connection. We can achieve something like this now on this library listening for the Stream `done`, but that's not simple. It would be a lot nicer if the library could provide a mechanism for auto retries.

**Comment by robertpiosik** [comment]
I'm not sure about this idea. [Node.js package with over 30M weekly downloads](https://www.npmjs.com/package/ws) doesn't provide any built-in mechanism for auto retires.

Wrapping connection in function which is calling itself with `Timer` in `onDone` handler is quite clean though.

**Comment by feinstein** [comment]
Well, IMHO we shouldn't guide ourselves on what others aren't doing.

That function will be very welcomed.

**Comment by robertpiosik** [comment]
It's just my experience - in real world app you likely do many things apart from simply reconnecting the connection. For instance, user is notified that the connection is lost and that requires from you dispatching some actions in order to set the app for such situation.

I believe that's why mentioned `retry` package is separate and not linked with `http` directly. It just handles exceptions from any async operation and http request happens to be one of them.

I'm not against your idea, I simply believe that `onDone` handler is sufficient.

**Comment by feinstein** [comment]
Yeah, I was thinking about a mechanism that can reconnect and have a callback to inform its reconnecting. I know it's simple, but if it comes from the library itself, then changes can be updated at once.

**Comment by PeterKnego** [comment]
I'd also welcome automatic reconnect logic. 

Using `onDone` handler is cumbersome because it's not reliable (it does not fire if app sleeps and socket meanwhile disconnects), and also it requires all higher logic to be rewired.

Ideally we would be able to set reconnect parameters at initialization and it would try to automatically reconnect within this parameters (num of retries, duration between retries, etc..).

At the moment to have a reliable permanent websocket connection we need to have an app-level ping message (low level ping messages are not exposed) which requires WS server to replies with pings, and then if connection goes down, we need to create new WebSocketChannel and rewire it app into logic.
